### DISCUSSION TITLE
popAndPush and autoDispose

### DISCUSSION THREAD

**Discussion by Ruzo** [rrousselGit/riverpod#743] [Q&A] [answered]
Hi, I'm trying to rebuild a ConsumerStatefulWidget while passing new parameters to it and resetting the state.

The widget:
```dart
class SearchPage extends ConsumerStatefulWidget {
  final String query;
  const SearchPage({this.query = '', Key? key}) : super(key: key);
```

auto_route router function - SearchRoute => SearchPage:
```dart
AutoRouter.of(context).popAndPush(
            SearchRoute(query: searchTerm),
          );
```

The provider:
```dart
final searchNotifierProvider = StateNotifierProvider.autoDispose<SearchNotifier, SearchState>(
  (ref) => SearchNotifier(
    ref.watch(searchRepositoryProvider),
  ),
);
```

I was hoping that popAndPush would force the provider to autoDispose but it doesn't. Even widget.query keeps it's previous data. I'm pretty sure it's a simple thing that I missed :)

**Comment by TimWhiting** [comment]
You can either refresh the provider: (Old Riverpod `context.refresh(searchNotifierProvider)` or dev version of Riverpod `ref.refresh(searchNotifierProvider)`) - likely in initState or didUpdateWidget. Or you can make the provider a family provider with the query string being the parameter for the family.

**Comment by rrousselGit** [comment]
Providers won't be disposed if they are immediately reused

So as Tim mentioned, you'd need to call "ref.refresh"

**Comment by Ruzo** [comment]
Thanks a lot for your quick replies, guys. It's good to know that ref.refresh is the right way to reset my state. However, I think the issue must be with `auto_route` because the `query` parameter is not receiving the new `searchTerm` in
```dart
AutoRouter.of(context).popAndPush(
            SearchRoute(query: searchTerm),
          );
```
The `SearchPage` rebuilds with the same value for `query`. I've added `maintainState: false` for `SearchRoute` but it still behaves the same way. Since I'm following a course, I'll stop taking your time and wait to see how Reso Coder deals with this feature.

Thank you again for your time.

**Reply by deepu** [reply]
Which course of Resocoder is this ?

**Reply by Ruzo** [reply]
https://resocoder.com/fdb/

**Comment by Ruzo** [comment]
I found a way to refresh the provider (instead of reloading the widget) and rerun the search function with the new searchTerm.
```dart
ref.refresh<SearchState>(searchNotifierProvider);
ref.read(searchNotifierProvider.notifier).getNextSearchedPage(currentTerm);
```
Thanks again.

**Comment by Ruzo** [comment, marked as answer]
By the way, the issue with popping and re-pushing the page was that I had to use `pushAndPopUntil(...)` instead of `popAndPush(...)`. So, `ref.refresh` then `pushAndPopUntil` works well.
